cosmos/config.py

```diff
@@ -46,39 +46,67 @@ class ProjectConfig:
     """
     Class for setting project config.
 
-    :param dbt_project_path: The path to the dbt project directory. Example: /path/to/dbt/project
+    :param dbt_project_path: The path to the dbt project directory. Example: /path/to/dbt/project. Defaults to None
     :param models_relative_path: The relative path to the dbt models directory within the project. Defaults to models
     :param seeds_relative_path: The relative path to the dbt seeds directory within the project. Defaults to seeds
     :param snapshots_relative_path: The relative path to the dbt snapshots directory within the project. Defaults to
     snapshots
     :param manifest_path: The absolute path to the dbt manifest file. Defaults to None
+    :param project_name: Allows the user to define the project name.
+    Required if dbt_project_path is not defined. Defaults to the folder name of dbt_project_path.
     """
 
-    dbt_project_path: str | Path
+    dbt_project_path: str | Path | None = None
     models_relative_path: str | Path = "models"
     seeds_relative_path: str | Path = "seeds"
     snapshots_relative_path: str | Path = "snapshots"
     manifest_path: str | Path | None = None
 
+    parsed_dbt_project_path: Path | None = None
```

Comment:
It feels like this would you think if this was a property of the class, as opposed to an attribute?

We could cache it using:
```
if sys.version_info >= (3, 9):
    from functools import cache
else:
    from functools import lru_cache
```

Reply:
I have attempted to implement this based on my understanding, however, i found an annotation called cached_property which seems to achieve what we want and integrates more nicely with a dataclass.

Reply:
Sounds great!